File: Haskell/2020 Septiembre/02_Hecho/AVL.hs
{------------------------------------------------------------------------------
 - Student's name:
 -
 - Student's group:
 -----------------------------------------------------------------------------}

module AVL 
  ( 
    Weight
  , Capacity
  , AVL (..)
  , Bin
  , emptyBin
  , remainingCapacity
  , addObject
  , maxRemainingCapacity
  , height
  , nodeWithHeight
  , node
  , rotateLeft
  , addNewBin
  , addFirst
  , addAll
  , toList
  , linearBinPacking
  , seqToList
  , addAllFold
  ) where

type Capacity = Int
type Weight= Int

data Bin = B Capacity [Weight] 

data AVL = Empty | Node Bin Int Capacity AVL AVL deriving Show


emptyBin :: Capacity -> Bin
emptyBin tam = (B tam [])

remainingCapacity :: Bin -> Capacity
remainingCapacity (B tam cap) = tam

addObject :: Weight -> Bin -> Bin
addObject o (B tam cap) 
  | o > tam = error "El cubo no cabe"
  | otherwise = (B (tam-o) (cap++[o]))

maxRemainingCapacity :: AVL -> Capacity
maxRemainingCapacity Empty = 0
maxRemainingCapacity (Node bin altura cap izq der) = cap

height :: AVL -> Int
height Empty = 0
height (Node bin altura cap izq der) = altura

 
nodeWithHeight :: Bin -> Int -> AVL -> AVL -> AVL
nodeWithHeight (B cap obj) alt Empty Empty = (Node (B cap obj) alt cap Empty Empty) 
nodeWithHeight (B cap obj) alt Empty (Node bind altd capd izqd derd)
  | capd > cap = (Node (B cap obj) alt capd Empty (Node bind altd capd izqd derd)) 
  | otherwise = (Node (B cap obj) alt cap Empty (Node bind altd capd izqd derd)) 
nodeWithHeight (B cap obj) alt (Node bini alti capi izqi deri) Empty
  | capi > cap = (Node (B cap obj) alt capi (Node bini alti capi izqi deri) Empty) 
  | otherwise = (Node (B cap obj) alt cap (Node bini alti capi izqi deri) Empty) 
nodeWithHeight (B cap obj) alt (Node bini alti capi izqi deri) (Node bind altd capd izqd derd)
  | capi >= capd && capi >= cap = (Node (B cap obj) alt capi (Node bini alti capi izqi deri) (Node bind altd capd izqd derd)) 
  | capd >= capi && capd >= cap = (Node (B cap obj) alt capd (Node bini alti capi izqi deri) (Node bind altd capd izqd derd)) 
  | otherwise = (Node (B cap obj) alt capi (Node bini alti cap izqi deri) (Node bind altd capd izqd derd)) 


node :: Bin -> AVL -> AVL -> AVL
node (B cap obj) Empty Empty = (Node (B cap obj) 1 cap Empty Empty)
node (B cap obj) Empty (Node bind altd capd izqd derd)
  | capd > cap = (Node (B cap obj) (altd+1) capd Empty (Node bind altd capd izqd derd)) 
  | otherwise = (Node (B cap obj) (altd+1) cap Empty (Node bind altd capd izqd derd)) 
node (B cap obj) (Node bini alti capi izqi deri) Empty
  | capi > cap = (Node (B cap obj) (alti+1) capi (Node bini alti capi izqi deri) Empty) 
  | otherwise = (Node (B cap obj) (alti+1) cap (Node bini alti capi izqi deri) Empty) 
node (B cap obj) (Node bini alti capi izqi deri) (Node bind altd capd izqd derd)
  | capi >= capd && capi >= cap = (Node (B cap obj) altura capi (Node bini alti capi izqi deri) (Node bind altd capd izqd derd)) 
  | capd >= capi && capd >= cap = (Node (B cap obj) altura capd (Node bini alti capi izqi deri) (Node bind altd capd izqd derd)) 
  | otherwise = (Node (B cap obj) altura capi (Node bini alti cap izqi deri) (Node bind altd capd izqd derd)) 
    where
      altura = maximaAltura alti altd

maximaAltura :: Int -> Int -> Int
maximaAltura x y
  | x > y = x
  | otherwise = y



rotateLeft :: Bin -> AVL -> AVL -> AVL
rotateLeft (B cap obj) izq (Node bind altd capd izqd derd) = x
  where
    c = node (B cap obj) izq izqd
    x = nodeWithHeight bind altd c derd


addNewBin :: Bin -> AVL -> AVL
addNewBin (B cap obj) Empty = (Node (B cap obj) 1 cap Empty Empty)
addNewBin (B cap obj) (Node (B cap1 obj1) alt capres izq der)
  | (height der) - (height izq)  > 1 = rotateLeft (B cap1 obj1) izq (addNewBin (B cap obj) der)
  | otherwise = (Node (B cap1 obj1) alt capres izq (addNewBin (B cap obj) der))

 
addFirst :: Capacity -> Weight -> AVL -> AVL
addFirst w x Empty = addNewBin (B (w-x) [x]) Empty
addFirst w x (Node bin alt cap izq der)
  | w < x = addNewBin (B (w-x) [x]) (Node bin alt cap izq der)
  | maxRemainingCapacity izq >= x = (Node bin alt cap (addFirst w x izq) der)
  | cap >= x = (Node (addObject x bin ) alt (cap-x) izq der)
  | otherwise = (Node bin alt cap izq (addFirst w x der))


addAll:: Capacity -> [Weight] -> AVL
addAll w [] = Empty
addAll w xs = addAllRec w xs Empty 

addAllRec :: Capacity -> [Weight] -> AVL -> AVL
addAllRec w [] sol = sol
addAllRec w (x:xs) sol = addAllRec w xs (addFirst w x sol)



toList :: AVL -> [Bin]
toList Empty = []
toList (Node bin alt cap izq der) = toList izq ++ [bin] ++ toList der 

{-
	SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
 -}

data Sequence = SEmpty | SNode Bin Sequence deriving Show  

linearBinPacking:: Capacity -> [Weight] -> Sequence
linearBinPacking _ _ = undefined

seqToList:: Sequence -> [Bin]
seqToList _ = undefined

addAllFold:: [Weight] -> Capacity -> AVL 
addAllFold _ _ = undefined



{- No modificar. Do not edit -}

objects :: Bin -> [Weight]
objects (B _ os) = reverse os

  
instance Show Bin where
  show b@(B c os) = "Bin("++show c++","++show (objects b)++")"

File: Java/2020 Septiembre/02_Hecho/src/AVL.java
/**
 * Student's name:
 *
 * Student's group:
 */

import dataStructures.list.ArrayList;
import dataStructures.list.List;


class Bin {
    private int remainingCapacity; // capacity left for this bin
    private List<Integer> weights; // weights of objects included in this bin

    public Bin(int initialCapacity) {
        // todo

        remainingCapacity = initialCapacity;
        weights = new ArrayList<>();
    }

    // returns capacity left for this bin
    public int remainingCapacity() {
        // todo
        return remainingCapacity;
    }

    // adds a new object to this bin
    public void addObject(int weight) {
        // todo

       /*if(weight > remainingCapacity)
        {
            throw new RuntimeException("El objeto no cabe en el cubo");
        }*/

        weights.append(weight);
        remainingCapacity = remainingCapacity - weight;

    }

    // returns an iterable through weights of objects included in this bin
    public Iterable<Integer> objects() {
        // todo
        //  SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
        //  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
        return null;
    }

    public String toString() {
        String className = getClass().getSimpleName();
        StringBuilder sb = new StringBuilder(className);
        sb.append("(");
        sb.append(remainingCapacity);
        sb.append(", ");
        sb.append(weights.toString());
        sb.append(")");
        return sb.toString();
    }
}

// Class for representing an AVL tree of bins
public class AVL {
    static private class Node {
        Bin bin; // Bin stored in this node
        int height; // height of this node in AVL tree
        int maxRemainingCapacity; // max capacity left among all bins in tree rooted at this node
        Node left, right; // left and right children of this node in AVL tree

        // recomputes height of this node
        void setHeight() {
            // todo
            if(left == null && right == null)
            {
                height = 1;
            } else if (left == null && right != null)
            {
                height = right.height+1;
            } else if (left != null && right == null)
            {
                height = left.height+1;
            } else
            {
                if(left.height > right.height)
                {
                    height = left.height+1;
                } else
                {
                    height = right.height+1;
                }
            }
        }

        // recomputes max capacity among bins in tree rooted at this node
        void setMaxRemainingCapacity() {
            // todo

            if(left == null && right == null)
            {
                maxRemainingCapacity = bin.remainingCapacity();
            } else if (left == null && right != null)
            {
              maxRemainingCapacity = Math.max(bin.remainingCapacity(), right.maxRemainingCapacity);
            } else if (left != null && right == null)
            {
                maxRemainingCapacity = Math.max(bin.remainingCapacity(), left.maxRemainingCapacity);
            } else
            {
                maxRemainingCapacity = Math.max(bin.remainingCapacity(), Math.max(left.maxRemainingCapacity, right.maxRemainingCapacity));
            }

        }

        // left-rotates this node. Returns root of resulting rotated tree
        Node rotateLeft() {
            // todo

            Node aux = this.right;
            this.right = aux.left;

            this.setMaxRemainingCapacity();
            this.setHeight();

            aux.left = this;

            aux.setMaxRemainingCapacity();
            aux.setHeight();

            return aux;
        }
    }

    private static int height(Node node) {
        // todo
        return node.height;
    }

    private static int maxRemainingCapacity(Node node) {
        // todo
        return node.maxRemainingCapacity;
    }

    private Node root; // root of AVL tree

    public AVL() {
        this.root = null;
    }

    // adds a new bin at the end of right spine.
    private void addNewBin(Bin bin) {
        // todo
        root = addNewBinRec(root,bin);
    }

    private Node addNewBinRec(Node nodo, Bin bin) {

        if(nodo == null)
        {
            Node nuevoNodo = new Node();
            nuevoNodo.bin = bin;
            nuevoNodo.setHeight();
            nuevoNodo.setMaxRemainingCapacity();
            nodo = nuevoNodo;
        } else
        {
            nodo.right = addNewBinRec(nodo.right, bin);
        }

        nodo.setHeight();
        nodo.setMaxRemainingCapacity();

        if(nodo.left != null && nodo.right != null)
        {
            if(nodo.left.height+1 < nodo.right.height)
            {
                nodo.rotateLeft();
            }
        }

        nodo.setHeight();
        nodo.setMaxRemainingCapacity();

        return nodo;

    }

    // adds an object to first suitable bin. Adds
    // a new bin if object cannot be inserted in any existing bin
    public void addFirst(int initialCapacity, int weight) {
        // todo

        addFirstRec(root, initialCapacity, weight);
    }

    private void addFirstRec(Node nodo, int initialCapacity, int weight) {

        if(nodo == null || nodo.maxRemainingCapacity < weight)
        {
            Bin b = new Bin(initialCapacity);
            b.addObject(weight);
            addNewBin(b);
        } else if (nodo.left != null && nodo.left.maxRemainingCapacity >= weight)
        {
            addFirstRec(nodo.left, initialCapacity, weight);
        } else if(nodo.bin.remainingCapacity() >= weight)
        {
            nodo.bin.addObject(weight);
        } else
        {
            addFirstRec(nodo.right, initialCapacity, weight);
        }
    }

    public void addAll(int initialCapacity, int[] weights) {
        // todo

        for(int i = 0; i<weights.length; i++)
        {
            addFirst(initialCapacity, weights[i]);
        }

    }



    public List<Bin> toList() {
        // todo

        List<Bin> l = new ArrayList<>();

        toListRec(root, l);

        return l;
    }

    private void toListRec(Node nodo, List<Bin> l) {

        if(nodo != null)
        {

            if(nodo.left != null)
            {
                toListRec(nodo.left,l);
            }

            l.append(nodo.bin);

            if(nodo.right != null)
            {
                toListRec(nodo.right, l);
            }
        }



    }

    public String toString() {
        String className = getClass().getSimpleName();
        StringBuilder sb = new StringBuilder(className);
        sb.append("(");
        stringBuild(sb, root);
        sb.append(")");
        return sb.toString();
    }

    private static void stringBuild(StringBuilder sb, Node node) {
        if(node==null)
            sb.append("null");
        else {
            sb.append(node.getClass().getSimpleName());
            sb.append("(");
            sb.append(node.bin);
            sb.append(", ");
            sb.append(node.height);
            sb.append(", ");
            sb.append(node.maxRemainingCapacity);
            sb.append(", ");
            stringBuild(sb, node.left);
            sb.append(", ");
            stringBuild(sb, node.right);
            sb.append(")");
        }
    }
}

class LinearBinPacking {
    public static List<Bin> linearBinPacking(int initialCapacity, List<Integer> weights) {
        // todo
        //  SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
        //  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
        return null;
    }
	
	public static Iterable<Integer> allWeights(Iterable<Bin> bins) {
        // todo
        //  SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
        //  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
        return null;		
	}
}